### PR TITLE
Allow debugging output in the testing library.

### PR DESCRIPTION
**Pull request by abizjak**
## Purpose

Enable debugging support in the contract testing library.

## Changes

The main change from the user's perspective is the addition of `debug_trace` in Invoke and Init results and errors, together with the `DebugInfoExt` trait which has utility functions for extracting and printing the debug information.

The testing library always runs the execution engine with debug output collection enabled to avoid additional generics.

I deem this performance impact acceptable due to the way the testing library is intended to be used.

Depends on
- [x] https://github.com/Concordium/concordium-base/pull/488
- [x] https://github.com/Concordium/concordium-rust-sdk/pull/147

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: examples/icecream/tests/tests.rs [hunk 1]
```diff
@@ -38,6 +38,7 @@ fn test_sunny_day() {
                     .expect("Serialize account address."),
             },
         )
+        .print_emitted_events()
```
Comment (limemloh) [inline examples/icecream/tests/tests.rs:41]: **Suggestion (non-blocking):** Maybe add a comment to what this line does, for educational purpose.

**Review by limemloh — APPROVED** [review]

File: concordium-std/src/lib.rs [hunk 2]
```diff
@@ -139,6 +140,21 @@
...
+//! During testing and debugging it is often useful to emit debug information to
+//! narrow down the source of the problem. `concordium-std` supports this using
+//! the [`concordium_dbg`] macro which will emit its arguments using a special
+//! host function `debug_print` which is only available when the `debug` feature
+//! is enabled. The output of this function is used by `cargo concordium run`
+//! and `cargo concordium test` to display any output that was emitted.
+//!
+//! The `debug` feature should typically not be enabled manually. It is used
+//! implicitly by `cargo concordium` when debug output is requested. It is also
+//! **crucial** that the `debug` feature is **not** enabled when building the
```
[the remark this answers is not in the document]
Comment (abizjak) [reply concordium-std/src/lib.rs:154]: Added

File: concordium-std/src/prims.rs [hunk 3]
```diff
@@ -282,6 +282,22 @@ extern "C" {
         column: u32,
     );
 
+    #[cfg(feature = "debug")]
+    /// Emit text together with the source location.
+    /// This is used as the equivalent of the `dbg!` macro when the
+    /// `enable-debug` feature is enabled.
```
[the remark this answers is not in the document]
Comment (abizjak) [reply concordium-std/src/prims.rs:288]: Fixed

File: contract-testing/src/invocation/impls.rs [hunk 4]
```diff
@@ -1269,58 +1353,68 @@ impl<'a, 'b> EntrypointInvocationHandler<'a, 'b> {
...
-            self.remaining_energy.tick_energy(used_energy)?;
-            Ok(to_interpreter_energy(*self.remaining_energy).energy)
-        };
+        let mut subtract_then_convert =
+            |mut remaining_energy| -> Result<InterpreterEnergy, InsufficientEnergy> {
+                // Using `saturating_sub` here should be ok since we should never be able to use
+                // more energy than what is available.
+                let used_energy = from_interpreter_energy(
+                    &InterpreterEnergy::new(available_interpreter_energy)
+                        .saturating_sub(&remaining_energy), /* TODO: This is stupid. Revise
```
[the remark this answers is not in the document]
Comment (abizjak) [reply contract-testing/src/invocation/impls.rs:1394]: I just removed the todo.

The "stupidity" was worse in some intermediate version of this.

File: examples/icecream/tests/tests.rs (hunk 1 above)
Comment (abizjak) [reply examples/icecream/tests/tests.rs]: Added

**Pull request merged**